### ISSUE TITLE
Any plans for Intellij Plugin Development

### ISSUE THREAD

**Issue by nagkumar** [TNG/ArchUnit#942]
I was presenting ArchUnit in Bangalore JUG today, and one of the attendees did give a great idea. Reporting the same for considering in future roadmaps

Is there a possibility of building the IDE plugin around #ArchUnit Rules, so that the rules violations can be caught during the code writing in IDE itself rather than the current way of building and running test cases to know the voilations?

**Comment by hankem** [comment]
One advantage of ArchUnit is that its tests can be executed just like any other unit test.

For instance, if you use plain JUnit `@Test`s (like in [`ArchUnit-Examples`'s `example-plain`](https://github.com/TNG/ArchUnit-Examples/tree/main/example-plain/src/test/java/com/tngtech/archunit/exampletest)), you can directly execute them individually from IntelliJ:

![image](https://user-images.githubusercontent.com/4208552/184480563-ccf67af7-cb35-452a-9619-05eae37d6947.png)

If you use [ArchUnit's JUnit support](https://www.archunit.org/userguide/html/000_Index.html#_junit_support) for `@ArchTest`s, i.e. the `ArchUnitRunner` for JUnit 4 or the `archunit-junit5-engine` for JUnit 5, IntelliJ supports running the entire test class file:

![image](https://user-images.githubusercontent.com/4208552/184480864-fea83926-7f26-4897-bc86-39122a898d9b.png)

**Comment by codecholeric** [comment]
This depends on what you want. As @hankem pointed out, it's easy to run the tests from the IDE like any ordinary unit test. This will compile the classes and then the bytecode can be imported. To have this e.g. as a live plugin within the IDE we would likely need to do a massive overhaul to allow filling the model from a different source. E.g. IntelliJ's PSI model. There once was an issue, but I never found the time for such a massive overhaul, since it would really mean designing new interfaces for users to plug into the import and also change the accessibility of the model completely (as it e.g. at the moment is not possible to create new `JavaClasses` for the users of ArchUnit and I think this is i.g. a good thing) -> #116
Otherwise, maybe it would also be possible to always compile the classes on the fly and import them, but I'm not sure how feasible that is :thinking: Given also, that ArchUnit at the moment allows no update of the class graph once it's created and again, changing that would likely be a highly non-trivial thing :thinking:

**Comment by nagkumar** [comment]
Agreed. I wish to give some more info on deeper IDE integration.

I can see the potential of archunit with proper libraries more than unit tests, such as helping developers learn faster the framework they are coding in.

With IDE deeper integration few  advantages to support developers are

1. Create ArchUnit 'different architecture' template test code by default. If we say onion architecture etc
For e.g when the spring project is created, also create respective arch unit source tests cases that are applicable to that framework
3. When Analysing the code, show deviations in IDE as in this image

![image](https://user-images.githubusercontent.com/332234/185741834-ab00dd2b-ebb0-48ec-88c8-a920db703d97.png)

With my current understanding of ArchUnit, I see it as Architecture as CODE (hence wish every framework provide these arch unit test library default, including Spring and in cloud space Amzaon Well-Architected framework etc), hence at least 50% to 70% of the architecture rules can be standardized and enforced as usual in JUnit test cases depending on the project type itself. This way learning curve of the new developers to that framework etc.. would drastically increase. I am sure this way, they can get used to unit tests too easily.

However, I do understand the bandwidth needed and the level of effort needed to make this happen. 

Thank you for your efforts in making archunit.

**Comment by codecholeric** [comment]
I completely agree, that it would be great to have ArchRules within your IDE as you type with direct feedback :slightly_smiling_face: I'm just afraid that it is really a big endeavor to get there with some real challenges to overcome. Even just thinking about the fact that I might want to evaluate rules on the fly that are part of my sources of my project. I.e. the plugin that checks the progress of me writing code depends on the compiled output of other code within the same project. I wouldn't know if IntelliJ e.g. would even allow that. And what if you would edit the rules themselves? But configured the rules you're editing to be checked against the code you're editing at the moment :zany_face: Just some things that come to mind, why it would at least be challenging (maybe impossible) to use the same patterns as for unit tests for a IDE plugin :shrug:
